### ISSUE TITLE
Nextcloud 12 (11) + Active Directory + SSO (Kerberos)

### ISSUE THREAD

**Issue by tkdimm** [nextcloud/user_saml#118]
Hello!

This description of the SSO configuration (Kerberos) and the problems.

**1. Enable apps "LDAP user and group backend" and "SSO & SAML authentication"**

**2. LDAP user and group backend**
Among other settings, on the Expert-tab need to change default attribute (UUID) on **userPrincipalName**

**3. SSO & SAML authentication**
Choose “Authentication via Environment Variable” and specify a value **REMOTE_USER**

**4. Apache**
If you set the SSO point to path:
<Location "**/index.php/login**">
AuthType Kerberos
…
</Location>

then at the point “/index.php/login” we get the next
**$_SERVER[‘REMOTE_USER’] = samaccountname@DOMAIN.LOCAL** (*Feature1)
Note that postfix userPrincipalName by default in uppercase.

**Problem1:**
$_SERVER[‘REMOTE_USER’] has is set value only in this point and unfortunately, it is not transmitted further (to user-saml scripts).

The solution of the problem may be a change of point:
<Location “**/index.php/apps/user_saml/saml/login**">
AuthType Kerberos
…
</Location>
This will allow the script to get the value $_SERVER[‘REMOTE_USER’]

**Problem2:**
In the database (**oc_accounts.uid**) value is written in lowercase: samaccountname@domain.local
Because of this (considering *Feature1), compliance is not met, therefore the input fails.

The solution to the problem can be achieved by adding a row:
File: _**/apps/user_saml/lib/Controller/SAMLController.php**_
_public function login() {
	…
case 'environment-variable':
$ssoUrl = $this->urlGenerator->getAbsoluteURL('/');
**// To lower $_SERVER[‘REMOTE_USER’]
$_SERVER['REMOTE_USER'] = strtolower($_SERVER['REMOTE_USER']);**
$this->session->set('user_saml.samlUserData', $_SERVER);
	…
}_

The proposed ways of solving these problems allow the work of SSO, but this "solution on the knee".

**Comment by jh23453** [comment]
Problem 1:
This most likely needs a change to the admin manual - I just yesterday looked into SSO and got to the same conclusion after some debugging. I'll try to prepare a pull request.

Problem 2:
I worked around it by using only the local username, without @domain and @REALM, because I was not successful in converting REMOTE_USER to lowercase.

In the Kerberos configuration for Apache I used:
```
    AuthType Kerberos
...
    KrbLocalUserMapping On

In the LDAP connection I used the following filter:
```
(&(|(objectclass=...))(|(uid=%uid)(|(mailPrimaryAddress=%uid)(mail=%uid))))
```
and in the expert tab uid as internal username.

**Comment by tkdimm** [comment]
Thank you for participating!

Unfortunately, it is not always possible to rely on the uniqueness of the uid.
For example, the case where several domains are the source.
There is a risk of opening a user profile with the same uid but from another domain.

Concerning the part of the request (|(mailPrimaryAddress=%uid)(mail=%uid))
Entering the uid in the primary email address is not always achievable.
For example, in many companies, there is a practice when changing the surname (marriage) does not change the user's uid (keep history of events and trace the life cycle account), but the alias of the mailbox is created and becomes the primary (not equal uid) and is changing DisplayName.
There is a risk of zero result when the filter is based on the e-mail address.

Thus, it is optimal to filter on userPrincipalName, but now it requires a local intervention in the user-saml to ensure operability with SSO... and these changes will be lost on the next update.

**Comment by jh23453** [comment]
For multiple/larger Domains your concerns are valid - my local net is really small and has a small userbase.

Another way to manipulate the REALM/domain might be mod_lookup_identity
https://www.adelton.com/apache/mod_lookup_identity/

**Comment by Erik-NA** [comment]
Is there any detailed documentation how to configure LDAP query in LDAP app to use SSO (Kerberos) with SAML app and Windows AD using Environment Variable?

**Comment by marcelin** [comment]
Hi everyone!!!

I've tried all ur configuration but SSO does't work. I'm using Nextcloud with "LDAP user and group backend" and "SSO & SAML authentication" enabled. i'm using windows server 2012 as LDAP auth service. My web sever is apache2

Please I need Help.

**Comment by bluikko** [comment]
I am evaluating migrating from OwnCloud to NextCloud to gain Kerberos SSO support (through mod_auth_gssapi on Apache and REMOTE_USER).

Could someone clarify does Kerberos SSO work or not?

In the OP:
* "Problem 1": What is the impact of this? Does this mean that Kerberos SSO does not work?
* "Problem 2": I could live with patching one line on a file manually after every update if I absolutely have to. But how come this issue has been open for 6 months with such a simple fix not committed? Are there some deeper issues with this feature?

**Comment by jh23453** [comment]
Problem 1: 
This seems to be (still) a problem with the documentation.
There is already a pull request pending to fix it:
https://github.com/nextcloud/documentation/pull/500

After that Kerberos SSO will work.

Problem 2:
It's no that easy in all cases - there might be users needing uppercase, so the mapping
would be wrong. And nextcould needs to work for most situations out-of-the-box.

I've since moved on to evaluating saml login with keycloak, which seems to work fine.

**Comment by bluikko** [comment]
@jh23453 

Re "Problem 2": Isn't that an invalid argument? If the user is saved into database always after converting to lower case, then case-sensitive mixed case login will always fail, not only with Kerberos?

**Comment by bluikko** [comment]
In Nextcloud 14 Administration Manual the whole section about "REMOTE_USER" has been removed. Does that mean the feature is totally removed even with the workarounds described in this issue?

**Comment by robertpavlovic10** [comment]
@bluikko And what is the situation of Kerberos SSO in Nextcloud 15?

**Comment by bluikko** [comment]
I did not try NC15. The whole project is now on hold and we might stay indefinitely with OwnCloud. The main reason to migrate from OC to NC was support for Kerberos. Migrating hundreds of users just do not seem worth it.

But at least on NC14 I did not have problem2 in OP (case mismatch in case-sensitive username realm part).
Problem1 in OP also seemed different but I do not know if there was something wrong with my configuration. I did not get it working with any other `<Location>` than `/` - both `/index.php/login` and `/index.php/apps/user_saml/saml/login` did not work but I do not remember anymore if I may have had (other) problem with saml_user configuration at that point.

All the different problems collectively put me off since even if the `<Location>` was fixed I would not have a usable system.

**Comment by Ornias1993** [comment]
@bluikko Just for documentation sake:
Nextcloud put all the "enterprise" documentation including SSO behind a paywalll...

**Comment by jospoortvliet** [comment]
Hi all,

If you're interested in moving a large deployment to Nextcloud, please consider working with the experts in our engineering team. The problems mentioned can easily be fixed by them in a PoC. This is exactly what we are here for! 

With regards to the issues reported here, this is more of a support request which we are happy to see reported through https://portal.nextcloud.com - our support team will help you debug and fix the problem following your SLA. If you have no SLA yet, get in contact with sales via our [website form](https://nextcloud.com/enterprise/buy/).

Github is primarily meant for developer collaboration and we're very happy with bug reports, but support tickets do not belong here. 

Just to clarify - a bug report is a clear description of a bug in Nextcloud, including the steps to reproduce it on a fresh Nextcloud developer instance. In simple terms, you have to do the work to identify the root cause so our developers can fix it right away. Example code, curl examples or example apps are good ways of showing what is wrong, and we very much welcome patches and links to the relevant code. Also, please make sure you have tested if the issue also happens on the latest daily build of Nextcloud. You can learn how to set up a developer instance here: https://docs.nextcloud.com/server/latest/developer_manual/

Of course, if you can't do this debugging, help.nextcloud.com has other home users who might be willing to help or give pointers. If your business need this issue fixed, we're happy to do the debugging for you as part of our subscription services.

I will close this now, feel free to re-open if you have debugged the problem and can provide more information.

**Comment by bluikko** [comment]
I guess that is one way of "resolving" a ticket - move the incomplete/incorrect documentation inside paywall and push more paid services.

**Comment by Ornias1993** [comment]
@bluikko and @jospoortvliet 
While I understand they are hesitant to provide support for enterprise features.
Putting the basic documentation behind a paywall, also makes it impossible to test said features.

With testing I mean both OpenSource development testing AND testing by interested potential enterprise customers.

We wouldn't need a PoC to make sure a feature works, if there was actually decent documentation available. I would call this a bug: A layer 7 bug, but a bug none the less.

Simply put: This Issue wouldn't have been opened if your documentation was in order. Hence its a bug. Although this bug shouldn't have been posted here, but posted here instead:
https://github.com/nextcloud/documentation

**Comment by Ornias1993** [comment]
@bluikko I found when it was removed and it seems suspicious at best.
Created an issue about it in the correct repo:
https://github.com/nextcloud/documentation/issues/1857

**Comment by omulki** [comment]
Hi folks.
Has there been any progress on this?
I can't seem to get SSO to work.
Apache is correctly configured, as is NC. I can even login via the basic-auth fallback functionality in Apache and everything works a treat. However, there is no SSO.

Thanks a lot and greeting, Spotlight.
